Do not forward HTCP CLR requests denied by htcp_clr_access

rousskov opened the pull request:
Before this change:

* htcp_access controls TST access
* htcp_clr_access controls CLR access to local cache purging
* nothing controls CLR forwarding -- it is always allowed

After this change:

* htcp_access controls TST access
* htcp_clr_access controls CLR access to local cache purging
* htcp_clr_access controls CLR forwarding

This bug was discovered and detailed by Joshua Rogers
https://joshua.hu/

rousskov commented:
If this PR changes are acceptable in principle, I will add a release note about them.

rousskov commented:
@MegaManSec, please update the line at the bottom of this PR description if you would like to add your email address or affiliation. Thank you for discovering this bug.

yadij reviewed (changes requested):
> TODO: implement the above; rename htcp_clr_access to htcp_purge_access.

Please do not. The "CLR" is very clearly HTCP feature documented in RFC specification. The word "PURGE" is burdened by implications of the deprecated custom Squid HTTP method.

> htcp_access controls TST access

Instead of documenting this for only about the TST packet (raising questions of why no `htcp_tst_access`) I would prefer clarify that the NOP, MON, SET packet types are ignored / not implemented.

IMO we should add an ACL type for HTCP operation. And move the `htcp_access` test into `htcpHandleMsg` before the packet-type switch. It would then truly apply to all HTCP types.

`htcp_clr_access` then becomes allow/deny about whether the packet is forwarded. Not whether it is applied locally.

rousskov commented:
> > TODO: implement the above; rename htcp_clr_access to htcp_purge_access.
> 
> Please do not.

I removed that TODO and associated text to avoid arguing about it.

> Instead of documenting this for only about the TST packet (raising questions of why no `htcp_tst_access`) I would prefer clarify that the NOP, MON, SET packet types are ignored / not implemented.
 
I added info about other HTCP opcodes.

I also added a release note.

> Or perhapse better;
> 
> Add an ACL type for HTCP operation, and move the `htcp_access` test into `htcpHandleMsg` before the packet-type switch. It would then truly apply to all HTCP types.

Sounds good in principle. FWIW, I believe we have far more important things to work on than HTCP improvements. I am only trying to close a security hole here, and hope to avoid doing even more HTCP work in the foreseeable future.

> `htcp_clr_access` then becomes allow/deny about whether the packet is forwarded. Not whether it is applied locally.

In that case, that directive should be renamed.
